lib/declare_schema/model.rb

```diff
@@ -263,9 +263,13 @@ def _infer_fk_limit(foreign_key_column, reflection)
       # unlike the `primary_key` method, DOES NOT query the database to find the actual primary key in use right now
       # if no explicit primary key set, returns the _default_declared_primary_key
       def _declared_primary_key
-        if defined?(@primary_key)
+        if !defined?(@primary_key) ||
+           (ActiveSupport.version >= Gem::Version.new('7.1.0') &&
+             @primary_key == ActiveRecord::AttributeMethods::PrimaryKey::ClassMethods::PRIMARY_KEY_NOT_SET)
```

Comment:
📓 In Rails 7.1+ they're now using a constant `BasicObject` instance as a reference to the primary key not being set yet, so this additional check is now necessary.